### PR TITLE
Implement core/widgets public API

### PR DESCRIPTION
**Pull request by ryanwelcher**
## Summary

Addresses issue #1301 

## Relevant technical choices

This PR branches PR #1550 done for #1298 which will need to be merged before this.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
@ryanwelcher it seems the build has broken (possibly from merging develop?) and there are a few merge conflicts to resolve before this one is ready for review.

**Comment by ryanwelcher** [comment]
@aaemnnosttv this is ready for CR!

File: assets/js/googlesitekit/widgets/index.js [hunk 1]
```diff
@@ -19,10 +19,102 @@
...
+	 * @param {string}             slug               Widget Area's slug.
+	 * @param {Object}             settings           Widget Area's settings.
+	 * @param {number}             settings.priority  Priority for this widget area. Default: 10.
+	 * @param {string}             settings.title     Title for this widget area.
+	 * @param {string}             settings.subtitle  Subtitle for this widget area.
+	 * @param {(string|undefined)} settings.icon      URL to SVG icon for this widget area.
+	 * @param {string}             settings.style     Widget area style (one of "boxes", "composite").
+	 * @param {(string|Array)}     areas              Widget Area slug(s).
+	 */
+	registerWidgetArea( slug, settings, areas ) {
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:42]: Nit-picking, but can we name the last parameter `areaSlugs` like in the datastore?

File: assets/js/googlesitekit/widgets/index.js [hunk 2]
```diff
@@ -19,10 +19,102 @@
...
+
+	/**
+	 * Register a widget.
+	 *
+	 * @since n.e.x.t
+	 * @param {string}         slug      Widget's slug.
+	 * @param {Object}         settings  Widget's settings.
+	 * @param {(string|Array)} contexts  Widget Context's slug(s).
+	 */
+	registerWidget( slug, settings, contexts ) {
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:55]: See above, can we name the last parameter `contextSlugs` like in the datastore?

File: assets/js/googlesitekit/widgets/index.js [hunk 3]
```diff
@@ -19,10 +19,102 @@
...
+	 *
+	 * @since n.e.x.t
+	 * @param {string}             slug               Widget Area's slug.
+	 * @param {Object}             settings           Widget Area's settings.
+	 * @param {number}             settings.priority  Priority for this widget area. Default: 10.
+	 * @param {string}             settings.title     Title for this widget area.
+	 * @param {string}             settings.subtitle  Subtitle for this widget area.
+	 * @param {(string|undefined)} settings.icon      URL to SVG icon for this widget area.
+	 * @param {string}             settings.style     Widget area style (one of "boxes", "composite").
+	 * @param {(string|Array)}     areas              Widget Area slug(s).
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:40]: Nit-picking, but this should be "Widget area slug(s)." (no need to capitalize the "a", it's not a name).

Regarding the signature, can we make this parameter optional? Mark as optional via `[areaSlugs]`, start parameter description with "Optional.", and in the function body only call `assignWidget` if `areaSlugs` is provided.

File: assets/js/googlesitekit/widgets/index.js [hunk 4]
```diff
@@ -19,10 +19,102 @@
...
+		dispatch( STORE_NAME ).assignWidget( slug, areas );
+	},
+
+	/**
+	 * Register a widget.
+	 *
+	 * @since n.e.x.t
+	 * @param {string}         slug      Widget's slug.
+	 * @param {Object}         settings  Widget's settings.
+	 * @param {(string|Array)} contexts  Widget Context's slug(s).
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:53]: See above, let's make this parameter optional, mark it as such and only call `assignWidgetArea` if it is provided. And let's lowercase the "context" :)

**Comment by felixarntz** [comment]
@aaemnnosttv 
> I noticed the API here is different than that defined in the AC, although it is consistent with the IB. Is this okay/expected?

We diverged from the original ACs during IB definition, so yeah let's follow the IB here. One thing not clearly stated there though is that these extra 3rd parameters should be optional, as in you should be able to register a widget area or widget without immediately assigning it. Left comments around that in my above CR. cc @ryanwelcher

File: assets/js/googlesitekit/widgets/index.js [hunk 5]
```diff
@@ -19,10 +19,102 @@
...
+	 * Register a widget area
+	 *
+	 * @since n.e.x.t
+	 * @param {string}             slug               Widget Area's slug.
+	 * @param {Object}             settings           Widget Area's settings.
+	 * @param {number}             settings.priority  Priority for this widget area. Default: 10.
+	 * @param {string}             settings.title     Title for this widget area.
+	 * @param {string}             settings.subtitle  Subtitle for this widget area.
+	 * @param {(string|undefined)} settings.icon      URL to SVG icon for this widget area.
+	 * @param {string}             settings.style     Widget area style (one of "boxes", "composite").
```
[the remark this answers is not in the document]
Comment (ryanwelcher) [reply assets/js/googlesitekit/widgets/index.js:39]: Updated

File: assets/js/googlesitekit/widgets/index.js (hunk 1 above)
Comment (ryanwelcher) [reply assets/js/googlesitekit/widgets/index.js:42]: updated

File: assets/js/googlesitekit/widgets/index.js (hunk 2 above)
Comment (ryanwelcher) [reply assets/js/googlesitekit/widgets/index.js:55]: updated

File: assets/js/googlesitekit/widgets/index.js (hunk 3 above)
Comment (ryanwelcher) [reply assets/js/googlesitekit/widgets/index.js:40]: updated

File: assets/js/googlesitekit/widgets/index.js (hunk 4 above)
Comment (ryanwelcher) [reply assets/js/googlesitekit/widgets/index.js:53]: updated

File: assets/js/googlesitekit/widgets/index.js [hunk 6]
```diff
@@ -19,10 +19,106 @@
...
+	 * Register a widget area
+	 *
+	 * @since n.e.x.t
+	 * @param {string}             slug                Widget Area's slug.
+	 * @param {Object}             settings            Widget Area's settings.
+	 * @param {string}             settings.title      Title for this widget area.
+	 * @param {string}             settings.subtitle   Subtitle for this widget area.
+	 * @param {number}             [settings.priority] Priority for this widget area. Default: 10.
+	 * @param {(string|undefined)} [settings.icon]     URL to SVG icon for this widget area.
+	 * @param {string}             [settings.style]    Widget area style (one of "boxes", "composite").
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:39]:
```suggestion
	 * @param {number}             [settings.priority] Optional. Priority for this widget area. Default: 10.
	 * @param {(string|undefined)} [settings.icon]     Optional. URL to SVG icon for this widget area.
	 * @param {string}             [settings.style]    Optional. Widget area style (one of "boxes", "composite"). Default: "boxes".
```

File: assets/js/googlesitekit/widgets/index.js [hunk 7]
```diff
@@ -19,10 +19,106 @@
...
+			dispatch( STORE_NAME ).assignWidget( slug, areaSlugs );
+		}
+	},
+
+	/**
+	 * Register a widget.
+	 *
+	 * @since n.e.x.t
+	 * @param {string}         slug           Widget's slug.
+	 * @param {Object}         settings       Widget's settings.
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:54]: The `settings` arguments should be documented here as well, similar like in the datastore and `registerWidgetArea` above:
```suggestion
	 * @param {string}          slug               Widget's slug.
	 * @param {Object}          settings           Widget's settings.
	 * @param {React.Component} settings.component React component used to display the contents of this widget.
	 * @param {number}          settings.priority  Optional. Widget's priority for ordering (lower number is higher priority, like WordPress hooks). Default is: 10.
	 * @param {string}          settings.width     Optional. Widget's maximum width to occupy. Default is: "quarter". One of: "quarter", "half", "full".
```

File: assets/js/googlesitekit/widgets/index.js [hunk 8]
```diff
@@ -19,10 +19,110 @@
...
+	 * Registers a widget.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}          slug               Widget's slug.
+	 * @param {Object}          settings           Widget's settings.
+	 * @param {React.Component} settings.component React component used to display the contents of this widget.
+	 * @param {number}          settings.priority  Optional. Widget's priority for ordering (lower number is higher priority, like WordPress hooks). Default is: 10.
+	 * @param {string}          settings.width     Optional. Widget's maximum width to occupy. Default is: "quarter". One of: "quarter", "half", "full".
+	 * @param {(string|Array)} [contextSlugs] Optional. Widget context slug(s).
```
Comment (felixarntz) [inline assets/js/googlesitekit/widgets/index.js:60]:
```suggestion
	 * @param {(string|Array)}  [contextSlugs]     Optional. Widget context slug(s).
```

**Pull request merged**